test/stdlib/StringDeconstruction.swift

```diff
@@ -0,0 +1,120 @@
+
+    switch expectDeconstruct {
+      case .scratchIfAvailable:
+        expectNil(ownerWS, stackTrace: stackTrace)
+        expectNotNil(ownerN, stackTrace: stackTrace)
+
+        expectEqual(scratch.baseAddress, rawBytesWS.baseAddress,
+          stackTrace: stackTrace)
+        expectNotEqual(scratch.baseAddress, rawBytesN.baseAddress,
+          stackTrace: stackTrace)
+
+        expectTrue(lengthWS < scratch.count, stackTrace: stackTrace)
+        expectTrue(lengthN < scratch.count, stackTrace: stackTrace)
+
+        expectTrue(usedScratchWS, stackTrace: stackTrace)
+        expectFalse(usedScratchN, stackTrace: stackTrace)
+
+        expectFalse(allocatedMemoryWS, stackTrace: stackTrace)
+        expectTrue(allocatedMemoryN, stackTrace: stackTrace)
+
+      case .interiorPointer:
+        // TODO: owner == (immortal ? nil : StringObject.largeAddress)
+        expectTrue(str.isContiguousUTF8, stackTrace: stackTrace)
+        var copy = str
+        copy.withUTF8 {
+          expectEqual($0.baseAddress, ptrWS, stackTrace: stackTrace)
+          expectEqual($0.baseAddress, ptrN, stackTrace: stackTrace)
+          expectEqual($0.count, lengthWS, stackTrace: stackTrace)
+          expectEqual($0.count, lengthN, stackTrace: stackTrace)
+        }
+
+        expectFalse(usedScratchWS, stackTrace: stackTrace)
+        expectFalse(usedScratchN, stackTrace: stackTrace)
+        expectFalse(allocatedMemoryWS, stackTrace: stackTrace)
+        expectFalse(allocatedMemoryN, stackTrace: stackTrace)
+      case .extraAllocation:
+        expectFalse(str.isContiguousUTF8, stackTrace: stackTrace)
+        expectNotNil(ownerWS, stackTrace: stackTrace)
+        expectNotNil(ownerN, stackTrace: stackTrace)
+        expectFalse(usedScratchWS, stackTrace: stackTrace)
+        expectFalse(usedScratchN, stackTrace: stackTrace)
+        expectTrue(allocatedMemoryWS, stackTrace: stackTrace)
+        expectTrue(allocatedMemoryN, stackTrace: stackTrace)
+    }
+  }
+}
+
+@inline(never)
+func id<T>(_ a: T) -> T { a }
+
+StringDeconstructTests.test("deconstruct") {
+  let smallASCII = "abcd"
+  let smallUTF8 = "ジッパー"
+  let large = "the quick fox jumped over the lazy brown dog"
+
+  var largeMortal = large
+  largeMortal.append(id("🧟‍♀️"))
+  largeMortal.append(id(largeMortal.last!))
+
+  expectDeconstruct(smallASCII, .scratchIfAvailable)
+  expectDeconstruct(smallUTF8, .scratchIfAvailable)
+  expectDeconstruct(large, .interiorPointer)
+  expectDeconstruct(largeMortal, .interiorPointer)
+}
+
+#if _runtime(_ObjC)
+import Foundation
+StringDeconstructTests.test("deconstruct cocoa") {
+  let smallCocoa: NSString = "aaa"
+  let largeASCIICocoa: NSString = "the quick fox jumped over the lazy brown dog"
+  let largeCocoa: NSString = "the quick 🧟‍♀️ ate the slow 🧠"
+
+  expectDeconstruct(smallCocoa as String, .scratchIfAvailable)
+  expectDeconstruct(largeASCIICocoa as String, .interiorPointer)
+  expectDeconstruct(largeCocoa as String, .extraAllocation)
+}
```

Comment:
@Catfish-Man we'd like to exhaustively test different possibilities of tagged, large contiguous ASCII, and large UTF-16 NSStrings. Is this the best way and/or is this complete? Is there something more guaranteed?

Reply:
Check out setUpStringTests in ObjectiveCTests.m for every possibility I could remember